src/recovery/mod.rs

```diff
@@ -563,8 +652,250 @@ impl RecoveryStore<u64, TdPyAny, TdPyAny> for SqliteRecoveryStore {
     }
 }
 
+/// A recovery store which save state to a Kafka topic.
+///
+/// We use the topic as a log of changes to a key-value store. We read
+/// back in the change log to re-build that state.
+///
+/// See [`KafkaKey`] for the kinds of things we want to store.
+struct KafkaRecoveryStore {
+    rt: SendWrapper<Runtime>,
+    hosts: Vec<String>,
+    producer: FutureProducer,
+    topic: String,
+}
+
+/// Keys in our Kafka-based key-value store.
+///
+/// Again use an enum to get easy serialization to / from bytes (which
+/// is all Kafka can actually read / write).
+#[derive(Debug, Serialize, Deserialize)]
+enum KafkaKey {
+    /// Store a state update for a stateful operator.
+    State {
+        step_id: StepId,
+        key: TdPyAny,
+        epoch: u64,
+    },
+    /// Save that the dataflow frontier has progressed.
+    DataflowFrontier,
+}
+
+impl KafkaKey {
+    fn to_bytes(&self) -> Vec<u8> {
+        // TODO: Figure out if there's a more robust-to-evolution way
+        // to serialize this key. If the serialization changes between
+        // versions, then recovery doesn't work. Or if we use an
+        // encoding that isn't deterministic.
+        bincode::serialize(self).expect("Error serializing state key")
+    }
+
+    fn from_bytes(bytes: &[u8]) -> Self {
+        bincode::deserialize(bytes).expect("Error deserializing state key")
+    }
+}
+
+/// Values our or Kafka-based key-value store.
+///
+/// Again use an enum to get easy serialization to / from bytes (which
+/// is all Kafka can actually read / write).
+#[derive(Debug, Serialize, Deserialize)]
+enum KafkaPayload {
+    /// The actual state update. If inside is `None` then the state
+    /// was deleted at this epoch.
+    State(Option<TdPyAny>),
+    /// A dataflow frontier value update.
+    DataflowFrontier(u64),
+}
+
+impl KafkaPayload {
+    fn to_bytes(&self) -> Vec<u8> {
+        bincode::serialize(self).expect("Error serializing state payload")
```

Comment:
I'm not too concerned. This level of schema is tied to the schema of our "recovery store DB", which will only change when we change the semantics of recovery (e.g. like perhaps not recovering on epoch boundaries in the future). I don't think we'll be doing that enough to need a sort of "on line" recovery technique. My gut stays an acceptable way forward here is that since recovery already means the dataflow is stopped, we could make bulk migration scripts that you can run on your recovery topic or SQLite DB to convert it to a new version, then restart.

But that does point out one difficulty that changes in that recovery state schema are multiplied across every recovery store implementation: you'll have to migrate the data in this Kafka topic and in the SQLite DB and ... all in the same way. Perhaps when we get to that we can find a way to make a generic transformer that works on all types.

I think we need a concrete case here to know what direction to go, but I don't think we've painted us into a corner yet. I'm more concerned about how we will evolve the dataflow logic state across restarts and if we need to move off Pickling at that level.